### DISCUSSION TITLE
AddChild() in increments

### DISCUSSION THREAD

**Discussion by ondesic** [godotengine/godot-proposals#6844] [2D]
I have been very grateful for the ability to create a thread and load scenes in the background with ResourceLoader Interactive.  However, once the scene is loaded, using ChangeSceneTo() or AddChild causes a freeze in the system.  This causes all animation to stop while it is working on adding the scene if it is a fairly large scene.
I have asked in other discussions if there is a way to AddChild() in another thread or asynchronously. No one was able to help.  
I noticed that some people will breakdown the large scene and add each child with CallDeferred().
I would love to have a function that can AddChild() (or ChangeSceneTo) like the Interactive resourceloader can.  Any thoughts?

**Comment by Calinou** [comment]
See https://github.com/godotengine/godot-proposals/issues/1801.

**Reply by ondesic** [reply]
@Calinou,
Does instancing a packed scene take the same time as AddChild() or ChangeSceneTo()?  Is this what is slowing down and freezing the animation.

**Reply by Zireael07** [reply]
You need to add_child an instanced packed scene anyway.

BTW I strongly suspect your freeze is shader compiling

**Reply by ondesic** [reply]
The program has no shaders.  It is a massive program and the main tscn scene that is being added (after the splash scene) is 15 MB. That is huge compared to the other scenes being added.
I would love way to incrementally add the scene in a way that frees up the main thread.

**Reply by Calinou** [reply]
Out of curiosity, why is your main scene so large? Shouldn't it be broken up in smaller parts, and save large resources to binary `.res` to speed up loading? (Don't keep ArrayMeshes in `.tscn`/`.tres` files.)

**Reply by theraot** [reply]
@ondesic
Would it be loading textures to GPU? Textures is something that could be making the scene file large.
Addendum: Audio could also be making it large. And yes, you are probably better off keeping those resources out of the scene file.

**Comment by ondesic** [comment]
My tscn file could be improved for sure.  It is the work of over 3 years and a massive program, however, I have seen others asking for a way to AddChild() in an incremental way as well.  It would be great to have an additional AddChildInteractive() function that dissects a large scene and adds each part in a way that allows an animation to play smoothly (without freezing) until the scene is added.

**Reply by theraot** [reply]
Adding it all at once makes sense, since the node tree is already built.

Anyway, yes, what you suggest could be possible. It must be in core since it would require to modify the `_ready` mechanism so it waits for pending children.

I wonder if something like `PackedScene.instantiate_to(Node)` would be preferable. Since the instantiation process will be building the node tree anyway, it could do it where the scene will be added to the scene tree directly, instead of having it built to then be migrated node by node to the scene tree. Edit: `instantiate_to` and `instantiate_to_interactive` (I don't know if `instantiate_to_threaded` is possible).

Yet, I suspect the issue are specific nodes taking more than a frame. And since the smaller unit you add to the scene remains a node, there would still be hiccups.
